### DISCUSSION TITLE
interrogate integral type of automatic result

### DISCUSSION THREAD

**Discussion by PaltryProgrammer** [boostorg/safe_numerics#136] [Ideas]
greetings kind regards
thank you for providing `safe_numerics` . i wish to know the integral type of a result of exempli gratia an addition of arguments with promotion policy `automatic` . upon examination of documentation and class definition i could find no such means . in particular i expected defined in class `safe_base` a public `using` to wit `using type = Stored;` i will modify the class to provide same and hope it works .
thank you kindly

**Comment by PaltryProgrammer** [comment]
`cout << typeid(safe_type<int>).name() << endl;`
results in 
`class boost::safe_numerics::safe_base<int,-2147483648,2147483647,struct boost::safe_numerics::automatic,struct boost::safe_numerics::exception_policy<struct throw_std_overflow_exception,struct throw_std_overflow_exception,struct throw_std_overflow_exception,struct boost::safe_numerics::ignore_exception> >`

i merely mish to know the integral type .

**Reply by robertramey** [reply]
consider something like:

template<typename boost::safe_numerics::safe<T> >
using underlying_type = T;

auto x = ....

cout << typeid(underlying_type<x>).name() << endl;

**Reply by PaltryProgrammer** [reply]
thank you kindly for reply
i do not see how said method can deduce underlying type of an `automatic` result .

as for `decltype` and `type_id` i can of course parse the resulting string but prefer to merely add `public: using type = Storred;` to `safe_base` which prvides a convenient means to obtain the enclosed type which every enclosing std class provides as far as i know .

having done so the result is as shown below . bingo presto ii seems to work .

input :
	safe_type<int> a = 159;
	safe_type<unsigned int> b = 0x159;
	auto z = a + b;
	cout << typeid(decltype(z)).name() << endl;
	cout << typeid(decltype(z)::type).name() << endl;

output :
class boost::safe_numerics::safe_base<__int64,-2147483648,6442450942,struct boost::safe_numerics::automatic,struct boost::safe_numerics::exception_policy<struct throw_std_overflow_exception,struct throw_std_overflow_exception,struct throw_std_overflow_exception,struct boost::safe_numerics::ignore_exception> >
__int64
